zlint: init at 3.4.0

baloo opened the pull request:
###### Description of changes

This adds `zlint` which is a linter for certificates.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

baloo committed:
zlint: init at 3.4.0

azahi reviewed (changes requested):
There were a lot of changes which I wanted to suggest so it would be easier if I just drop a patch for you to apply.

```diff
diff --git i/pkgs/tools/security/zlint/default.nix w/pkgs/tools/security/zlint/default.nix
index 4116e3bf717..9d85b181a62 100644
--- i/pkgs/tools/security/zlint/default.nix
+++ w/pkgs/tools/security/zlint/default.nix
@@ -1,4 +1,10 @@
-{ buildGoModule, fetchFromGitHub, lib }:
+{ lib
+, buildGoModule
+, fetchFromGitHub
+, git
+, testers
+, zlint
+}:
 
 buildGoModule rec {
   pname = "zlint";
@@ -6,25 +12,54 @@ buildGoModule rec {
 
   src = fetchFromGitHub {
     owner = "zmap";
-    repo = pname;
+    repo = "zlint";
     rev = "v${version}";
-    sha256 = "sha256-l39GdfEKUAw5DQNjx6ZBgfGtengRlUUasm0G07kAA2A=";
+    leaveDotGit = true;
+    hash = "sha256-1T8WAWsivSEB2xVEM+GpWJuD3DGXPa9uNpuN6/ABsns=";
   };
 
   modRoot = "v3";
+
   vendorHash = "sha256-OiHEyMHuSiWDB/1YRvAhErb1h/rFfXXVcagcP386doc=";
-  preBuild = ''
-    # not in the go.mod
-    rm -rf cmd/genTestCerts
+
+  postPatch = ''
+    # Remove a package which is not declared in go.mod.
+    rm -rf v3/cmd/genTestCerts
+  '';
+
+  subPackages = [
+    "cmd/zlint"
+    "cmd/zlint-gtld-update"
+  ];
+
+  ldflags = [
+    "-s"
+    "-w"
+    "-X main.version=${version}"
+  ];
+
+  checkInputs = [ git ];
+
+  preCheck = ''
+    # Test all targets.
+    unset subPackages
   '';
 
-  # Tests rely on git and we don't have the .git dir because modRoot is in a subdir
-  doCheck = false;
+  passthru.tests.version = testers.testVersion {
+    package = zlint;
+    command = "zlint -version";
+  };
 
   meta = with lib; {
-    homepage = "https://github.com/zmap/zlint/";
+    description = "X.509 Certificate Linter focused on Web PKI standards and requirements";
+    longDescription = ''
+      ZLint is a X.509 certificate linter written in Go that checks for
+      consistency with standards (e.g. RFC 5280) and other relevant PKI
+      requirements (e.g. CA/Browser Forum Baseline Requirements).
+    '';
+    homepage = "https://github.com/zmap/zlint";
+    changelog = "https://github.com/zmap/zlint/releases/tag/${src.rev}";
     license = licenses.asl20;
-    description = "X.509 Certificate Linter focused on Web PKI standards and requirements.";
     maintainers = with maintainers; [ baloo ];
   };
 }
```

Here's the complete [file](https://gist.github.com/azahi/e699c218930fa43514c8846fc0f04f98) if applying didn't work.

baloo commented:
@azahi that is a lot more than I know about! Would you want to be added in the maintainers maybe?

baloo committed:
zlint: apply review's comment

azahi commented:
The build currently [fails](https://logs.nix.ci/?key=nixos/nixpkgs.201537&attempt_id=ac4618c3-46cf-4762-8b4e-7bc6a08f9dc1) due to a hash mismatch, could you please fix that. Otherwise I think we're good.

> Would you want to be added in the maintainers maybe?

That's not necessary.

baloo commented:
@azahi the hash mismatch is inconsistent.
I got `fetchFromGitHub` to generate both. I believe the `leaveDotGit` renders the derivation non-reproducible.

baloo commented:
```
$ nix-build /nix/store/axhgyqdmpr9h3x4s0b4fsnnaa2kni1hz-source.drv --check
checking outputs of '/nix/store/axhgyqdmpr9h3x4s0b4fsnnaa2kni1hz-source.drv'...
exporting https://github.com/zmap/zlint.git (rev v3.4.0) into /nix/store/fp5p9nmpjn3k61df75dr37zrv3vnman9-source
Initialized empty Git repository in /nix/store/fp5p9nmpjn3k61df75dr37zrv3vnman9-source/.git/
remote: Enumerating objects: 1431, done.
remote: Counting objects: 100% (1431/1431), done.
remote: Compressing objects: 100% (1003/1003), done.
remote: Total 1431 (delta 652), reused 1039 (delta 422), pack-reused 0
Receiving objects: 100% (1431/1431), 2.00 MiB | 2.65 MiB/s, done.
Resolving deltas: 100% (652/652), done.
From https://github.com/zmap/zlint
 * tag               v3.4.0     -> FETCH_HEAD
Switched to a new branch 'fetchgit'
Enumerating objects: 1430, done.
Counting objects: 100% (1430/1430), done.
Compressing objects: 100% (1424/1424), done.
Writing objects: 100% (1430/1430), done.
Total 1430 (delta 623), reused 764 (delta 0), pack-reused 0
Enumerating objects: 1430, done.
Nothing new to pack.
/nix/store/fp5p9nmpjn3k61df75dr37zrv3vnman9-source
$ nix-build /nix/store/axhgyqdmpr9h3x4s0b4fsnnaa2kni1hz-source.drv --check
checking outputs of '/nix/store/axhgyqdmpr9h3x4s0b4fsnnaa2kni1hz-source.drv'...
exporting https://github.com/zmap/zlint.git (rev v3.4.0) into /nix/store/fp5p9nmpjn3k61df75dr37zrv3vnman9-source
Initialized empty Git repository in /nix/store/fp5p9nmpjn3k61df75dr37zrv3vnman9-source/.git/
remote: Enumerating objects: 1431, done.
remote: Counting objects: 100% (1431/1431), done.
remote: Compressing objects: 100% (1003/1003), done.
remote: Total 1431 (delta 652), reused 1040 (delta 422), pack-reused 0
Receiving objects: 100% (1431/1431), 2.00 MiB | 3.90 MiB/s, done.
Resolving deltas: 100% (652/652), done.
From https://github.com/zmap/zlint
 * tag               v3.4.0     -> FETCH_HEAD
Switched to a new branch 'fetchgit'
Enumerating objects: 1430, done.
Counting objects: 100% (1430/1430), done.
Compressing objects: 100% (1424/1424), done.
Writing objects: 100% (1430/1430), done.
Total 1430 (delta 623), reused 764 (delta 0), pack-reused 0
Enumerating objects: 1430, done.
Nothing new to pack.
error: hash mismatch in fixed-output derivation '/nix/store/axhgyqdmpr9h3x4s0b4fsnnaa2kni1hz-source.drv':
         specified: sha256-1T8WAWsivSEB2xVEM+GpWJuD3DGXPa9uNpuN6/ABsns=
            got:    sha256-se8Z31dBhoarqouFkVjOPIG86K5G/VabIKT9V6RgRKM=
```

baloo committed:
part ways with leaveDotGit

azahi reviewed (approved):
That's weird, but I guess we can just drop checks. LGTM.

cole-h reviewed (approved):
Seems fine to me.

Result of `nixpkgs-review pr 201537` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>zlint</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/tools/security/zlint/default.nix:
```diff
@@ -0,0 +1,59 @@
+    "-X main.version=${version}"
+  ];
+
+  # Checks rely on .git directory, leaveDotGit makes the source derivation flaky.
+  doCheck = false;
```
Can we patch that it requires a git directory?

baloo replied:
that got dropped in 3.4.1 which just released on Sunday and that I missed.
https://github.com/NixOS/nixpkgs/pull/202590
